File: dlitscustom/dlitscustom/report/dlits_tax_report/dlits_tax_report.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

frappe.query_reports["DLITS Tax Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "transaction_type",
            "label": __("Transaction Type"),
            "fieldtype": "Select",
            "options": "\nSales\nPurchase",
            "default": ""
        },
        {
            "fieldname": "tax_account",
            "label": __("Tax Account"),
            "fieldtype": "Link",
            "options": "Account",
            "get_query": function() {
                return {
                    "filters": {
                        "account_type": ["in", ["Tax", "Chargeable"]],
                        "is_group": 0
                    }
                };
            }
        },
        {
            "fieldname": "tax_rate",
            "label": __("Tax Rate (%)"),
            "fieldtype": "Float",
            "precision": 2
        },
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer"
        },
        {
            "fieldname": "supplier",
            "label": __("Supplier"),
            "fieldtype": "Link",
            "options": "Supplier"
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Format summary rows
        if (data && data.tax_account) {
            if (data.tax_account.includes('<b>') || data.tax_account.includes('Total')) {
                value = `<div style="font-weight: bold; color: #2e7d32;">${value}</div>`;
            }
            
            if (data.tax_account.includes('─')) {
                return `<div style="border-top: 2px solid #ddd; margin: 5px 0;"></div>`;
            }
        }
        
        // Color code transaction types
        if (column.fieldname === "transaction_type" && data) {
            if (data.transaction_type === "Sales") {
                value = `<span style="color: #28a745; font-weight: bold;">${value}</span>`;
            } else if (data.transaction_type === "Purchase") {
                value = `<span style="color: #dc3545; font-weight: bold;">${value}</span>`;
            }
        }
        
        // Format tax amounts
        if (column.fieldname === "tax_amount" && data && data.tax_amount) {
            if (data.transaction_type === "Sales") {
                value = `<span style="color: #28a745;">${value}</span>`;
            } else if (data.transaction_type === "Purchase") {
                value = `<span style="color: #dc3545;">${value}</span>`;
            }
        }
        
        return value;
    },
    
    "onload": function(report) {
        // Add custom buttons for export functionality
        report.page.add_inner_button(__("Export to PDF"), function() {
            let filters = report.get_values();
            let filter_string = get_filter_string(filters);
            
            frappe.call({
                method: "frappe.utils.print_format.download_pdf",
                args: {
                    doctype: "Sales Taxes and Charges",
                    name: "DLITS Tax Report",
                    format: "Standard",
                    doc: {
                        "name": "DLITS Tax Report",
                        "filters": filters,
                        "data": report.data || []
                    }
                },
                callback: function(r) {
                    if (r.message) {
                        // Create a more comprehensive PDF export
                        export_to_pdf(report, filters);
                    }
                }
            });
        }, __("Export"));
        
        report.page.add_inner_button(__("Export to Excel"), function() {
            let filters = report.get_values();
            export_to_excel(report, filters);
        }, __("Export"));
        
        // Add refresh button
        report.page.add_inner_button(__("Refresh"), function() {
            report.refresh();
        }, __("Actions"));
    }
};

function get_filter_string(filters) {
    let filter_parts = [];
    
    if (filters.from_date) filter_parts.push(`From: ${filters.from_date}`);
    if (filters.to_date) filter_parts.push(`To: ${filters.to_date}`);
    if (filters.transaction_type) filter_parts.push(`Type: ${filters.transaction_type}`);
    if (filters.tax_account) filter_parts.push(`Tax Account: ${filters.tax_account}`);
    if (filters.tax_rate) filter_parts.push(`Tax Rate: ${filters.tax_rate}%`);
    if (filters.customer) filter_parts.push(`Customer: ${filters.customer}`);
    if (filters.supplier) filter_parts.push(`Supplier: ${filters.supplier}`);
    
    return filter_parts.join(" | ");
}

function export_to_pdf(report, filters) {
    let data = report.data || [];
    let columns = report.columns || [];
    
    // Create HTML content for PDF
    let html_content = `
        <div style="text-align: center; margin-bottom: 30px;">
            <h1 style="color: #2e7d32; margin-bottom: 10px;">DLITS Tax Report</h1>
            <p style="color: #666; font-size: 14px;">${get_filter_string(filters)}</p>
            <p style="color: #666; font-size: 12px;">Generated on: ${frappe.datetime.now_datetime()}</p>
        </div>
        
        <table style="width: 100%; border-collapse: collapse; font-size: 12px;">
            <thead>
                <tr style="background-color: #f8f9fa;">
                    ${columns.map(col => `<th style="border: 1px solid #ddd; padding: 8px; text-align: left;">${col.label}</th>`).join('')}
                </tr>
            </thead>
            <tbody>
                ${data.map(row => `
                    <tr>
                        ${columns.map(col => {
                            let value = row[col.fieldname] || '';
                            if (col.fieldtype === 'Currency') {
                                value = format_currency(value);
                            }
                            return `<td style="border: 1px solid #ddd; padding: 6px;">${value}</td>`;
                        }).join('')}
                    </tr>
                `).join('')}
            </tbody>
        </table>
    `;
    
    // Create and download PDF
    let pdf_window = window.open('', '_blank');
    pdf_window.document.write(`
        <html>
            <head>
                <title>DLITS Tax Report</title>
                <style>
                    body { font-family: Arial, sans-serif; margin: 20px; }
                    table { page-break-inside: auto; }
                    tr { page-break-inside: avoid; page-break-after: auto; }
                </style>
            </head>
            <body>
                ${html_content}
                <script>
                    window.onload = function() {
                        window.print();
                        setTimeout(function() { window.close(); }, 1000);
                    };
                </script>
            </body>
        </html>
    `);
    pdf_window.document.close();
}

function export_to_excel(report, filters) {
    let data = report.data || [];
    let columns = report.columns || [];
    
    if (!data.length) {
        frappe.msgprint(__("No data to export"));
        return;
    }
    
    // Prepare data for Excel export
    let excel_data = [];
    
    // Add header with report title and filters
    excel_data.push(['DLITS Tax Report']);
    excel_data.push([get_filter_string(filters)]);
    excel_data.push([`Generated on: ${frappe.datetime.now_datetime()}`]);
    excel_data.push([]); // Empty row
    
    // Add column headers
    excel_data.push(columns.map(col => col.label));
    
    // Add data rows
    data.forEach(row => {
        let excel_row = columns.map(col => {
            let value = row[col.fieldname] || '';
            if (col.fieldtype === 'Currency' && value) {
                return parseFloat(value) || 0;
            }
            return value;
        });
        excel_data.push(excel_row);
    });
    
    // Create and download Excel file
    let worksheet = XLSX.utils.aoa_to_sheet(excel_data);
    let workbook = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(workbook, worksheet, "Tax Report");
    
    // Set column widths
    let col_widths = columns.map(col => ({wch: Math.max(col.label.length, 15)}));
    worksheet['!cols'] = col_widths;
    
    // Generate filename with timestamp
    let filename = `DLITS_Tax_Report_${frappe.datetime.now_date()}.xlsx`;
    
    // Download file
    XLSX.writeFile(workbook, filename);
    
    frappe.show_alert({
        message: __("Excel file downloaded successfully"),
        indicator: "green"
    });
}

function format_currency(value) {
    if (!value) return '0.00';
    return parseFloat(value).toLocaleString('en-US', {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2
    });
}

// Load XLSX library if not already loaded
if (typeof XLSX === 'undefined') {
    frappe.require('/assets/frappe/js/lib/xlsx.full.min.js');
}

File: dlitscustom/dlitscustom/report/dlits_tax_report/dlits_tax_report.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate, add_days
from datetime import datetime

def execute(filters=None):
    if not filters:
        filters = {}
    
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data, filters)
    
    return columns, data, None, chart

def get_columns():
    return [
        {
            "label": _("Tax Account"),
            "fieldname": "tax_account",
            "fieldtype": "Link",
            "options": "Account",
            "width": 200
        },
        {
            "label": _("Tax Rate (%)"),
            "fieldname": "tax_rate",
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "label": _("Transaction Type"),
            "fieldname": "transaction_type",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Document"),
            "fieldname": "document",
            "fieldtype": "Dynamic Link",
            "options": "document_type",
            "width": 150
        },
        {
            "label": _("Document Type"),
            "fieldname": "document_type",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Date"),
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 100
        },
        {
            "label": _("Customer/Supplier"),
            "fieldname": "party",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "label": _("Net Amount"),
            "fieldname": "net_amount",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Tax Amount"),
            "fieldname": "tax_amount",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Total Amount"),
            "fieldname": "total_amount",
            "fieldtype": "Currency",
            "width": 120
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    
    # Get Sales Invoice Tax Data
    sales_tax_data = frappe.db.sql("""
        SELECT 
            stc.account_head as tax_account,
            stc.rate as tax_rate,
            'Sales' as transaction_type,
            si.name as document,
            'Sales Invoice' as document_type,
            si.posting_date,
            si.customer as party,
            si.net_total as net_amount,
            stc.tax_amount,
            si.grand_total as total_amount
        FROM `tabSales Taxes and Charges` stc
        INNER JOIN `tabSales Invoice` si ON stc.parent = si.name
        WHERE si.docstatus = 1 {sales_conditions}
        ORDER BY si.posting_date DESC, stc.rate DESC
    """.format(sales_conditions=conditions.get('sales', '')), filters, as_dict=1)
    
    # Get Purchase Invoice Tax Data
    purchase_tax_data = frappe.db.sql("""
        SELECT 
            ptc.account_head as tax_account,
            ptc.rate as tax_rate,
            'Purchase' as transaction_type,
            pi.name as document,
            'Purchase Invoice' as document_type,
            pi.posting_date,
            pi.supplier as party,
            pi.net_total as net_amount,
            ptc.tax_amount,
            pi.grand_total as total_amount
        FROM `tabPurchase Taxes and Charges` ptc
        INNER JOIN `tabPurchase Invoice` pi ON ptc.parent = pi.name
        WHERE pi.docstatus = 1 {purchase_conditions}
        ORDER BY pi.posting_date DESC, ptc.rate DESC
    """.format(purchase_conditions=conditions.get('purchase', '')), filters, as_dict=1)
    
    # Combine and sort data
    all_data = sales_tax_data + purchase_tax_data
    
    # Sort by date (newest first) and then by tax rate
    all_data.sort(key=lambda x: (x.posting_date, x.tax_rate), reverse=True)
    
    # Add summary rows
    summary_data = get_summary_data(all_data)
    
    return summary_data + all_data

def get_conditions(filters):
    conditions = {'sales': '', 'purchase': ''}
    
    if filters.get("from_date"):
        conditions['sales'] += " AND si.posting_date >= %(from_date)s"
        conditions['purchase'] += " AND pi.posting_date >= %(from_date)s"
    
    if filters.get("to_date"):
        conditions['sales'] += " AND si.posting_date <= %(to_date)s"
        conditions['purchase'] += " AND pi.posting_date <= %(to_date)s"
    
    if filters.get("tax_account"):
        conditions['sales'] += " AND stc.account_head = %(tax_account)s"
        conditions['purchase'] += " AND ptc.account_head = %(tax_account)s"
    
    if filters.get("tax_rate"):
        conditions['sales'] += " AND stc.rate = %(tax_rate)s"
        conditions['purchase'] += " AND ptc.rate = %(tax_rate)s"
    
    if filters.get("transaction_type"):
        if filters.get("transaction_type") == "Sales":
            conditions['purchase'] = " AND 1=0"  # Exclude purchase data
        elif filters.get("transaction_type") == "Purchase":
            conditions['sales'] = " AND 1=0"  # Exclude sales data
    
    if filters.get("customer"):
        conditions['sales'] += " AND si.customer = %(customer)s"
    
    if filters.get("supplier"):
        conditions['purchase'] += " AND pi.supplier = %(supplier)s"
    
    return conditions

def get_summary_data(data):
    if not data:
        return []
    
    # Group by tax account and rate
    tax_summary = {}
    sales_total = 0
    purchase_total = 0
    
    for row in data:
        key = f"{row.tax_account}_{row.tax_rate}"
        if key not in tax_summary:
            tax_summary[key] = {
                'tax_account': row.tax_account,
                'tax_rate': row.tax_rate,
                'sales_amount': 0,
                'purchase_amount': 0,
                'net_amount': 0,
                'total_tax': 0
            }
        
        if row.transaction_type == 'Sales':
            tax_summary[key]['sales_amount'] += flt(row.tax_amount)
            sales_total += flt(row.tax_amount)
        else:
            tax_summary[key]['purchase_amount'] += flt(row.tax_amount)
            purchase_total += flt(row.tax_amount)
        
        tax_summary[key]['total_tax'] += flt(row.tax_amount)
    
    # Create summary rows
    summary_rows = []
    
    # Overall summary
    summary_rows.append({
        'tax_account': '<b>TOTAL TAX SUMMARY</b>',
        'tax_rate': '',
        'transaction_type': '',
        'document': '',
        'document_type': '',
        'posting_date': '',
        'party': '',
        'net_amount': '',
        'tax_amount': f'<b>{sales_total + purchase_total:.2f}</b>',
        'total_amount': ''
    })
    
    summary_rows.append({
        'tax_account': 'Sales Tax Total',
        'tax_rate': '',
        'transaction_type': 'Sales',
        'document': '',
        'document_type': '',
        'posting_date': '',
        'party': '',
        'net_amount': '',
        'tax_amount': f'{sales_total:.2f}',
        'total_amount': ''
    })
    
    summary_rows.append({
        'tax_account': 'Purchase Tax Total',
        'tax_rate': '',
        'transaction_type': 'Purchase',
        'document': '',
        'document_type': '',
        'posting_date': '',
        'party': '',
        'net_amount': '',
        'tax_amount': f'{purchase_total:.2f}',
        'total_amount': ''
    })
    
    # Add separator
    summary_rows.append({
        'tax_account': '─' * 50,
        'tax_rate': '',
        'transaction_type': '',
        'document': '',
        'document_type': '',
        'posting_date': '',
        'party': '',
        'net_amount': '',
        'tax_amount': '',
        'total_amount': ''
    })
    
    return summary_rows

def get_chart_data(data, filters):
    if not data:
        return None
    
    # Filter out summary rows for chart
    chart_data = [row for row in data if not (
        row.get('tax_account', '').startswith('<b>') or 
        row.get('tax_account', '').startswith('Sales Tax Total') or
        row.get('tax_account', '').startswith('Purchase Tax Total') or
        '─' in str(row.get('tax_account', ''))
    )]
    
    if not chart_data:
        return None
    
    # Group by transaction type
    sales_total = sum(flt(row.tax_amount) for row in chart_data if row.transaction_type == 'Sales')
    purchase_total = sum(flt(row.tax_amount) for row in chart_data if row.transaction_type == 'Purchase')
    
    return {
        "data": {
            "labels": ["Sales Tax", "Purchase Tax"],
            "datasets": [
                {
                    "name": "Tax Amount",
                    "values": [sales_total, purchase_total]
                }
            ]
        },
        "type": "donut",
        "height": 300,
        "colors": ["#28a745", "#dc3545"]
    }